File: nodejs/tests/convert.test.ts
/**
 * Turning generated messages into the SDK's public types.
 *
 * The conventions under test are the two the module documents: an empty string
 * becomes `null` where absence is meaningful, and a date the service sends is
 * kept as an ISO string only when it is genuinely one.
 */

import assert from 'node:assert/strict'
import { test } from 'node:test'

import { MemcoInternalError } from '../src/errors.js'
import {
  feedbackSubmitter,
  toDomainList,
  toFeedbackResult,
  toImportResult,
  toMemory,
  toRevertResult,
  toSearchResult,
  toSessionFields,
  toWriteResult,
  type FeedbackCaller
} from '../src/internal/convert.js'
import * as pb from '../src/internal/gen.js'
import {
  ImportStatus,
  RevertOutcome,
  type FeedbackRating
} from '../src/types.js'

/** A shareFeedback stub that fails the test if it is ever actually called. */
const NO_SHARE: FeedbackCaller = () => {
  throw new Error('shareFeedback was not expected to be called in this test')
}

/** A submitter that fails the test if `feedback()` is ever actually called. */
const NO_FEEDBACK = feedbackSubmitter('unused', NO_SHARE)

test('an instructions part with nothing to say stays an empty string', () => {
  // The contract documents an empty part as a real state, so it is not folded
  // to null the way an absent notice or operation id is.
  const session = toSessionFields(
    pb.StartSessionResponse.fromPartial({ sessionId: 's' })
  )
  assert.equal(session.instructions.content, '')
  assert.equal(session.instructions.policy, '')
})

test('a response carrying no instructions message reads as nothing to say', () => {
  // ts-proto leaves an unset nested message undefined rather than handing back
  // a default instance, so the conversion is what has to turn it into the empty
  // instructions a caller can read without a guard.
  const session = toSessionFields(
    pb.StartSessionResponse.fromPartial({ sessionId: 's' })
  )
  assert.deepEqual(session.instructions, {
    content: '',
    policy: '',
    adding: '',
    rating: '',
    next: ''
  })
})

test('instructions are read part for part', () => {
  const session = toSessionFields(
    pb.StartSessionResponse.fromPartial({
      sessionId: 's',
      instructions: {
        content: 'c',
        policy: 'p',
        adding: 'a',
        rating: 'r',
        next: 'n'
      }
    })
  )
  assert.deepEqual(session.instructions, {
    content: 'c',
    policy: 'p',
    adding: 'a',
    rating: 'r',
    next: 'n'
  })
})

test('repeated fields become plain arrays', () => {
  const result = toDomainList(
    pb.ListDomainsResponse.fromPartial({
      domains: [{ slug: 'coding', filterTagTypes: ['language'] }]
    })
  )
  assert.ok(Array.isArray(result.domains))
  assert.deepEqual(result.domains[0].filterTagTypes, ['language'])
})

test('a domain entry is read field for field', () => {
  const result = toDomainList(
    pb.ListDomainsResponse.fromPartial({
      domains: [
        {
          slug: 'coding',
          title: 'Coding',
          summary: 'How this team builds software',
          whenToSearch: 'before writing code',
          whenToSave: 'after learning something',
          whatNotToSave: 'secrets',
          tagsDescription: 'language, framework',
          filterTagTypes: ['language'],
          versionTagTypes: ['framework'],
          maxTagsPerQuery: 4
        }
      ]
    })
  )
  assert.deepEqual(result.domains[0], {
    slug: 'coding',
    title: 'Coding',
    summary: 'How this team builds software',
    whenToSearch: 'before writing code',
    whenToSave: 'after learning something',
    whatNotToSave: 'secrets',
    tagsDescription: 'language, framework',
    filterTagTypes: ['language'],
    versionTagTypes: ['framework'],
    maxTagsPerQuery: 4
  })
})

test('an insight update is kept as the ISO day the service sent', () => {
  // Not a Date: reading a date-only value as a timestamp attaches the reader's
  // time zone to it, landing the day early for anyone west of UTC.
  const memory = toMemory(
    pb.MemoryResult.fromPartial({
      idx: 'memory-a-1',
      insights: [{ updated: '2026-08-26' }]
    }),
    NO_FEEDBACK
  )
  assert.equal(memory.insights[0].updated, '2026-08-26')
})

for (const raw of [
  '',
  'not-a-date',
  '2026-13-99',
  '2026-02-30',
  '2025-02-29',
  '2026-8-26',
  '20260826',
  '2026-08-26T00:00:00Z',
  '2026-08-26 ',
  '2026-08-26\n',
  '+2026-08-26',
  '0000-01-01',
  '0000-02-29'
]) {
  test(`an update the service sent as ${JSON.stringify(raw)} becomes null`, () => {
    // A malformed date is never fatal: the rest of the insight is still useful.
    // Every value here is one that `new Date(raw)` would have accepted or
    // silently shifted, which is why the parse is hand-rolled.
    const memory = toMemory(
      pb.MemoryResult.fromPartial({ idx: 'm', insights: [{ updated: raw }] }),
      NO_FEEDBACK
    )
    assert.equal(memory.insights[0].updated, null)
  })
}

test('a leap day is a real date in a leap year', () => {
  const memory = toMemory(
    pb.MemoryResult.fromPartial({
      idx: 'm',
      insights: [{ updated: '2024-02-29' }]
    }),
    NO_FEEDBACK
  )
  assert.equal(memory.insights[0].updated, '2024-02-29')
})

test('empty optional strings become null', () => {
  const search = toSearchResult(
    pb.SearchResponse.fromPartial({ sessionId: 's', notice: '' }),
    NO_SHARE
  )
  assert.equal(search.notice, null)
  assert.equal(
    toMemory(pb.MemoryResult.fromPartial({ idx: 'm' }), NO_FEEDBACK).reference,
    null
  )
  const feedback = toFeedbackResult(
    pb.ShareFeedbackResponse.fromPartial({
      sessionId: 's',
      entries: [{ idx: 'i', advice: '' }]
    })
  )
  assert.equal(feedback.entries[0].advice, null)
})

test('populated optional strings are kept', () => {
  const search = toSearchResult(
    pb.SearchResponse.fromPartial({ sessionId: 's', notice: 'heads up' }),
    NO_SHARE
  )
  assert.equal(search.notice, 'heads up')
  const memory = toMemory(
    pb.MemoryResult.fromPartial({ idx: 'm', reference: 'memory-a-1' }),
    NO_FEEDBACK
  )
  assert.equal(memory.reference, 'memory-a-1')
})

test('an empty operation id marks a write that cannot be undone', () => {
  assert.equal(
    toWriteResult(pb.CreateMemoryResponse.fromPartial({ operationId: '' }))
      .operationId,
    null
  )
  assert.equal(
    toWriteResult(pb.CreateMemoryResponse.fromPartial({ operationId: 'op-1' }))
      .operationId,
    'op-1'
  )
})

test('a create response and an enrich response convert identically', () => {
  // The two messages are structurally the same, which is why one converter
  // serves both. If they ever diverge this is what says so.
  const created = toWriteResult(
    pb.CreateMemoryResponse.fromPartial({
      operationId: 'op-1',
      instructions: { next: 'revert with op-1' }
    })
  )
  const enriched = toWriteResult(
    pb.EnrichMemoryResponse.fromPartial({
      operationId: 'op-1',
      instructions: { next: 'revert with op-1' }
    })
  )
  assert.deepEqual(created, enriched)
})

test('an absent limits message means validate nothing, not validate zero', () => {
  // Reading an unset message as zeros would reject every call before it was
  // sent, locally, with no way for the caller to tell why.
  const result = toDomainList(
    pb.ListDomainsResponse.fromPartial({ domains: [{ slug: 'coding' }] })
  )
  assert.equal(result.limits, null)
})

test('a limits message is read field for field', () => {
  const result = toDomainList(
    pb.ListDomainsResponse.fromPartial({
      limits: {
        maxQueryCharacters: 100,
        maxTextCharacters: 200,
        maxIdxCharacters: 32,
        maxSources: 3,
        maxFeedbackEntries: 5,
        maxImportMemories: 50,
        maxImportQueriesPerMemory: 6,
        maxImportInsightsPerMemory: 7,
        maxImportTagsPerMemory: 8
      }
    })
  )
  assert.deepEqual(result.limits, {
    maxQueryCharacters: 100,
    maxTextCharacters: 200,
    maxIdxCharacters: 32,
    maxSources: 3,
    maxFeedbackEntries: 5,
    maxImportMemories: 50,
    maxImportQueriesPerMemory: 6,
    maxImportInsightsPerMemory: 7,
    maxImportTagsPerMemory: 8
  })
})

test('the deprecation fields reach the caller', () => {
  const result = toDomainList(
    pb.ListDomainsResponse.fromPartial({
      deprecated: true,
      deprecationMessage: 'Memory API v1 is superseded; migrate to v2.',
      sunsetDate: '2027-01-01'
    })
  )
  assert.equal(result.deprecated, true)
  assert.equal(
    result.deprecationMessage,
    'Memory API v1 is superseded; migrate to v2.'
  )
  assert.equal(result.sunsetDate, '2027-01-01')
})

test('an older service reporting no deprecation says nothing', () => {
  const result = toDomainList(pb.ListDomainsResponse.fromPartial({}))
  assert.equal(result.deprecated, false)
  assert.equal(result.deprecationMessage, '')
  assert.equal(result.sunsetDate, null)
})

test('the answering build reaches the caller', () => {
  // It names the build that answered, which is what a bug report quotes. It is
  // not provenance().serverCommit, which is the commit this package was cut at.
  assert.equal(
    toDomainList(
      pb.ListDomainsResponse.fromPartial({ serverCommit: '8317b7b' })
    ).serverCommit,
    '8317b7b'
  )
  assert.equal(
    toDomainList(pb.ListDomainsResponse.fromPartial({})).serverCommit,
    ''
  )
})

test('a revert carries the outcome as a typed value', () => {
  const result = toRevertResult(
    pb.RevertMemoryResponse.fromPartial({
      operationId: 'op-1',
      outcome: pb.RevertOutcome.REVERT_OUTCOME_MEMORY_REMOVED
    })
  )
  assert.equal(result.outcome, RevertOutcome.MEMORY_REMOVED)
  assert.equal(result.operationId, 'op-1')
})

test('a revert outcome this SDK predates folds to unspecified', () => {
  const result = toRevertResult(
    pb.RevertMemoryResponse.fromPartial({ outcome: 999 as pb.RevertOutcome })
  )
  assert.equal(result.outcome, RevertOutcome.UNSPECIFIED)
  assert.equal(result.operationId, null)
})

test('an import carries one typed outcome per entry', () => {
  // The indices are deliberately not their own positions: reading them off the
  // array index instead of off the field would pass an in-order fixture.
  const result = toImportResult([
    [
      0,
      pb.ImportMemoriesResponse.fromPartial({
        results: [
          { index: 2, status: pb.ImportStatus.IMPORT_STATUS_QUEUED },
          { index: 0, status: pb.ImportStatus.IMPORT_STATUS_DUPLICATE },
          {
            index: 1,
            status: pb.ImportStatus.IMPORT_STATUS_REJECTED,
            errors: ['queries must not be empty']
          }
        ]
      })
    ]
  ])
  assert.deepEqual(
    result.results.map(outcome => outcome.index),
    [2, 0, 1]
  )
  assert.equal(result.results[0].status, ImportStatus.QUEUED)
  assert.equal(result.results[1].status, ImportStatus.DUPLICATE)
  assert.deepEqual(result.results[2].errors, ['queries must not be empty'])
  // An entry that was queued carries an empty list, not a null.
  assert.deepEqual(result.results[0].errors, [])
})

test('an import status this SDK predates folds to unspecified', () => {
  const result = toImportResult([
    [
      0,
      pb.ImportMemoriesResponse.fromPartial({
        results: [{ index: 0, status: 999 as pb.ImportStatus }]
      })
    ]
  ])
  assert.equal(result.results[0].status, ImportStatus.UNSPECIFIED)
})

test('split groups are renumbered against the whole batch', () => {
  // Each call numbers its own results from zero. Merging them without the
  // offset would report index 0 once per group and identify nothing.
  const result = toImportResult([
    [
      0,
      pb.ImportMemoriesResponse.fromPartial({
        results: [
          { index: 0, status: pb.ImportStatus.IMPORT_STATUS_QUEUED },
          { index: 1, status: pb.ImportStatus.IMPORT_STATUS_QUEUED }
        ],
        instructions: { content: 'first' }
      })
    ],
    [
      2,
      pb.ImportMemoriesResponse.fromPartial({
        results: [
          { index: 0, status: pb.ImportStatus.IMPORT_STATUS_DUPLICATE },
          { index: 1, status: pb.ImportStatus.IMPORT_STATUS_QUEUED }
        ],
        instructions: { content: 'second' }
      })
    ]
  ])
  assert.deepEqual(
    result.results.map(outcome => outcome.index),
    [0, 1, 2, 3]
  )
  assert.equal(result.results[2].status, ImportStatus.DUPLICATE)
  // The calls are one operation in one domain, so one set of guidance is right.
  assert.equal(result.instructions.content, 'first')
})

test('a search result nests its memories and their insights', () => {
  const result = toSearchResult(
    pb.SearchResponse.fromPartial({
      sessionId: 'session-a',
      memories: [
        {
          idx: 'memory-a-1',
          kind: 'insight',
          timesServed: 3,
          intents: ['why does X happen'],
          insights: [
            {
              idx: 'memory-a-1-insight-1',
              title: 'T',
              content: 'C',
              updated: '2026-08-26',
              timesServed: 2,
              endorsed: 1,
              disputed: 0
            }
          ]
        }
      ]
    }),
    NO_SHARE
  )
  assert.equal(result.sessionId, 'session-a')
  const memory = result.memories[0]
  assert.equal(memory.idx, 'memory-a-1')
  assert.equal(memory.kind, 'insight')
  assert.equal(memory.timesServed, 3)
  assert.deepEqual(memory.intents, ['why does X happen'])
  assert.deepEqual(memory.insights[0], {
    idx: 'memory-a-1-insight-1',
    title: 'T',
    content: 'C',
    updated: '2026-08-26',
    timesServed: 2,
    endorsed: 1,
    disputed: 0
  })
})

test('a feedback result carries what the service recorded per rating', () => {
  const result = toFeedbackResult(
    pb.ShareFeedbackResponse.fromPartial({
      sessionId: 'session-a',
      entries: [
        {
          idx: 'memory-a-1',
          relevant: true,
          correct: false,
          advice: 'check it'
        }
      ]
    })
  )
  assert.equal(result.sessionId, 'session-a')
  assert.deepEqual(result.entries[0], {
    idx: 'memory-a-1',
    relevant: true,
    correct: false,
    advice: 'check it'
  })
})

test('a repeated field is copied, not aliased to the message', () => {
  // A caller holding the result must not see it change when the message it
  // came from is reused or recycled.
  const message = pb.ListDomainsResponse.fromPartial({
    domains: [{ slug: 'coding', filterTagTypes: ['language'] }]
  })
  const result = toDomainList(message)
  message.domains[0].filterTagTypes.push('framework')
  assert.deepEqual(result.domains[0].filterTagTypes, ['language'])
})

test('a sunset date the service malformed becomes null, not a throw', () => {
  // The notice is still worth showing without its date.
  const result = toDomainList(
    pb.ListDomainsResponse.fromPartial({
      deprecated: true,
      deprecationMessage: 'migrate to v2',
      sunsetDate: 'whenever'
    })
  )
  assert.equal(result.sunsetDate, null)
  assert.equal(result.deprecationMessage, 'migrate to v2')
})

// -- feedback wiring --------------------------------------------------------

test("a memory's feedback() calls the submitter with its own idx", async () => {
  const calls: [string, unknown][] = []
  const memory = toMemory(
    pb.MemoryResult.fromPartial({ idx: 'memory-a-1' }),
    async (idx, rating) => {
      calls.push([idx, rating])
      return {
        idx,
        relevant: rating.relevant,
        correct: rating.correct,
        advice: null
      }
    }
  )
  const entry = await memory.feedback({ relevant: true, correct: false })
  assert.deepEqual(calls, [['memory-a-1', { relevant: true, correct: false }]])
  assert.deepEqual(entry, {
    idx: 'memory-a-1',
    relevant: true,
    correct: false,
    advice: null
  })
})

test('a submitter forwards a comment to the caller unchanged', async () => {
  const received: unknown[] = []
  const submit = feedbackSubmitter('session-a', async options => {
    received.push([...options.feedback])
    return {
      sessionId: 'session-a',
      entries: [
        { idx: 'memory-a-1', relevant: true, correct: true, advice: null }
      ],
      instructions: {
        content: '',
        policy: '',
        adding: '',
        rating: '',
        next: ''
      }
    }
  })
  await submit('memory-a-1', { relevant: true, correct: true, comment: 'why' })
  assert.deepEqual(received, [
    [{ idx: 'memory-a-1', relevant: true, correct: true, comment: 'why' }]
  ])
})

test('a submitter throws a typed error rather than returning undefined when entries is empty', async () => {
  const submit = feedbackSubmitter('session-a', async () => ({
    sessionId: 'session-a',
    entries: [],
    instructions: {
      content: '',
      policy: '',
      adding: '',
      rating: '',
      next: ''
    }
  }))
  await assert.rejects(
    submit('memory-a-1', { relevant: true, correct: true }),
    (error: unknown) => {
      assert.ok(error instanceof MemcoInternalError)
      assert.match(error.detail, /no feedback entry.*"memory-a-1"/)
      return true
    }
  )
})

test('toSearchResult builds its submitter from the response own session id', async () => {
  const seen: Array<{
    sessionId: string
    rating: Pick<FeedbackRating, 'idx' | 'relevant' | 'correct'>
  }> = []
  const result = toSearchResult(
    pb.SearchResponse.fromPartial({
      sessionId: 'session-from-response',
      memories: [{ idx: 'memory-a-1' }, { idx: 'memory-a-2' }]
    }),
    async options => {
      const [rating] = [...options.feedback]
      seen.push({
        sessionId: options.sessionId,
        rating: {
          idx: rating!.idx,
          relevant: rating!.relevant,
          correct: rating!.correct
        }
      })
      return {
        sessionId: options.sessionId,
        entries: [{ idx: 'x', relevant: true, correct: true, advice: null }],
        instructions: {
          content: '',
          policy: '',
          adding: '',
          rating: '',
          next: ''
        }
      }
    }
  )
  await result.memories[0]!.feedback({ relevant: true, correct: true })
  await result.memories[1]!.feedback({ relevant: false, correct: false })
  // Every memory in the result shares the one submitter, built from the
  // response's own session id rather than one supplied by the caller — and
  // each call still carries its own memory's idx and its own rating, so a
  // regression where every memory closed over the same idx would fail here.
  assert.deepEqual(seen, [
    {
      sessionId: 'session-from-response',
      rating: { idx: 'memory-a-1', relevant: true, correct: true }
    },
    {
      sessionId: 'session-from-response',
      rating: { idx: 'memory-a-2', relevant: false, correct: false }
    }
  ])
})

File: go/memcoai/convert_test.go
package memcoai

import (
	"reflect"
	"testing"

	"google.golang.org/protobuf/proto"

	memoryv1 "github.com/memcoai/memcoai/go/internal/client/memcoai/memory/v1"
)

var wireInstructions = &memoryv1.Instructions{Content: "c", Policy: "p", Adding: "a", Rating: "r", Next: "n"}

func TestInstructionsAreReadPartForPart(t *testing.T) {
	if got := toInstructions(wireInstructions); got != (Instructions{"c", "p", "a", "r", "n"}) {
		t.Fatalf("got %+v", got)
	}
	if got := toInstructions(nil); got != (Instructions{}) {
		t.Fatalf("absent instructions: %+v", got)
	}
}

func TestADomainListIsReadFieldForField(t *testing.T) {
	response := &memoryv1.ListDomainsResponse{
		Domains: []*memoryv1.DomainEntry{{
			Slug: "coding", Title: "Coding", Summary: "s", WhenToSearch: "ws", WhenToSave: "wv",
			WhatNotToSave: "wn", TagsDescription: "td", FilterTagTypes: []string{"language"},
			VersionTagTypes: []string{"library"}, MaxTagsPerQuery: 5,
		}},
		Instructions: wireInstructions,
		Limits: &memoryv1.Limits{
			MaxQueryCharacters: 1, MaxTextCharacters: 2, MaxIdxCharacters: 3, MaxSources: 4,
			MaxFeedbackEntries: 5, MaxImportMemories: 6, MaxImportQueriesPerMemory: 7,
			MaxImportInsightsPerMemory: 8, MaxImportTagsPerMemory: 9,
		},
		Deprecated:         true,
		DeprecationMessage: "upgrade",
		SunsetDate:         "2027-01-01",
		ServerCommit:       "abc",
	}
	want := &DomainList{
		Domains: []DomainEntry{{
			Slug: "coding", Title: "Coding", Summary: "s", WhenToSearch: "ws", WhenToSave: "wv",
			WhatNotToSave: "wn", TagsDescription: "td", FilterTagTypes: []string{"language"},
			VersionTagTypes: []string{"library"}, MaxTagsPerQuery: 5,
		}},
		Instructions:       Instructions{"c", "p", "a", "r", "n"},
		Limits:             &Limits{1, 2, 3, 4, 5, 6, 7, 8, 9},
		Deprecated:         true,
		DeprecationMessage: "upgrade",
		SunsetDate:         "2027-01-01",
		ServerCommit:       "abc",
	}
	got := toDomainList(response)
	if !reflect.DeepEqual(got, want) {
		t.Fatalf("got %+v", got)
	}
	response.Domains[0].FilterTagTypes[0] = "changed"
	if got.Domains[0].FilterTagTypes[0] != "language" {
		t.Fatal("the result shares a slice with the response")
	}
}

func TestAnOlderServiceReportsNoLimitsAndNoNotice(t *testing.T) {
	got := toDomainList(&memoryv1.ListDomainsResponse{SunsetDate: "someday"})
	if got.Limits != nil || got.Deprecated || got.DeprecationMessage != "" || got.SunsetDate != "" {
		t.Fatalf("got %+v", got)
	}
	if got.Domains == nil {
		t.Fatal("an empty domain list reads as nil")
	}
}

func TestTheToolCatalogIsRead(t *testing.T) {
	got := toTools(&memoryv1.ListToolsResponse{Tools: []*memoryv1.ToolDescriptor{
		{Name: "search", Description: "d", Available: true},
		{Name: "revert_memory"},
	}})
	want := []ToolDescriptor{{"search", "d", true}, {"revert_memory", "", false}}
	if !reflect.DeepEqual(got, want) {
		t.Fatalf("got %+v", got)
	}
}

func wireMemory() *memoryv1.MemoryResult {
	return &memoryv1.MemoryResult{
		Idx: "memory-a-1", Kind: "policy", TimesServed: 3, Intents: []string{"why"},
		Insights: []*memoryv1.InsightResult{{
			Idx: "memory-a-1-insight-1", Title: "t", Content: "c", Updated: "2026-08-26",
			TimesServed: 2, Endorsed: 1, Disputed: 4,
		}, {
			Idx: "memory-a-1-insight-2", Updated: "2026-02-30",
		}},
	}
}

func TestAMemoryIsReadAndBound(t *testing.T) {
	ops := &MemoryOperations{}
	got := toMemory(wireMemory(), ops, "session-a")
	want := Memory{
		Idx: "memory-a-1", Kind: "policy", TimesServed: 3, Intents: []string{"why"},
		Insights: []Insight{
			{Idx: "memory-a-1-insight-1", Title: "t", Content: "c", Updated: "2026-08-26", TimesServed: 2, Endorsed: 1, Disputed: 4},
			{Idx: "memory-a-1-insight-2"},
		},
		ops:       ops,
		sessionID: "session-a",
	}
	if !reflect.DeepEqual(got, want) {
		t.Fatalf("got %+v", got)
	}
	reference := toMemory(&memoryv1.MemoryResult{Idx: "memory-b-1", Reference: "memory-a-1"}, ops, "s")
	if reference.Reference != "memory-a-1" || reference.Insights == nil || len(reference.Insights) != 0 {
		t.Fatalf("got %+v", reference)
	}
}

func TestASearchResultBindsItsOwnSession(t *testing.T) {
	ops := &MemoryOperations{}
	got := toSearchResult(&memoryv1.SearchResponse{
		SessionId: "session-new", Notice: "narrowed", Instructions: wireInstructions,
		Memories: []*memoryv1.MemoryResult{wireMemory()},
	}, ops)
	if got.SessionID != "session-new" || got.Notice != "narrowed" || got.Instructions.Content != "c" {
		t.Fatalf("got %+v", got)
	}
	if len(got.Memories) != 1 || got.Memories[0].sessionID != "session-new" || got.Memories[0].ops != ops {
		t.Fatalf("memories %+v", got.Memories)
	}
	if empty := toSearchResult(&memoryv1.SearchResponse{}, ops); empty.Memories == nil || empty.Notice != "" {
		t.Fatalf("empty %+v", empty)
	}
}

func TestWritesFeedbackAndRevertsAreRead(t *testing.T) {
	if got := toWriteResult("create-a", wireInstructions); got.OperationID != "create-a" || got.Instructions.Next != "n" {
		t.Fatalf("write %+v", got)
	}
	if got := toWriteResult("", nil); got.OperationID != "" || got.Instructions != (Instructions{}) {
		t.Fatalf("un-revertible write %+v", got)
	}
	feedback := toFeedbackResult(&memoryv1.ShareFeedbackResponse{
		SessionId: "s", Instructions: wireInstructions,
		Entries: []*memoryv1.FeedbackEntry{{Idx: "i", Relevant: true, Correct: false, Advice: "look again"}, {Idx: "j"}},
	})
	if !reflect.DeepEqual(feedback.Entries, []FeedbackEntry{{"i", true, false, "look again"}, {Idx: "j"}}) || feedback.SessionID != "s" {
		t.Fatalf("feedback %+v", feedback)
	}
	revert := toRevertResult(&memoryv1.RevertMemoryResponse{OperationId: "create-a", Outcome: memoryv1.RevertOutcome_REVERT_OUTCOME_EXPIRED})
	if revert.OperationID != "create-a" || revert.Outcome != RevertOutcomeExpired {
		t.Fatalf("revert %+v", revert)
	}
	if unknown := toRevertResult(&memoryv1.RevertMemoryResponse{Outcome: 42}); unknown.Outcome != RevertOutcomeUnspecified {
		t.Fatalf("unknown outcome %+v", unknown)
	}
}

func TestSplitImportsAreRenumberedToTheCallersPositions(t *testing.T) {
	got := toImportResult([]importPart{
		{0, &memoryv1.ImportMemoriesResponse{Instructions: wireInstructions, Results: []*memoryv1.ImportOutcome{
			{Index: 0, Status: memoryv1.ImportStatus_IMPORT_STATUS_QUEUED},
			{Index: 1, Status: memoryv1.ImportStatus_IMPORT_STATUS_REJECTED, Errors: []string{"too short"}},
		}}},
		{2, &memoryv1.ImportMemoriesResponse{Instructions: &memoryv1.Instructions{Content: "second"}, Results: []*memoryv1.ImportOutcome{
			{Index: 0, Status: 42},
		}}},
	})
	want := &ImportResult{
		Results: []ImportOutcome{
			{Index: 0, Status: ImportStatusQueued, Errors: []string{}},
			{Index: 1, Status: ImportStatusRejected, Errors: []string{"too short"}},
			{Index: 2, Status: ImportStatusUnspecified, Errors: []string{}},
		},
		Instructions: Instructions{"c", "p", "a", "r", "n"},
	}
	if !reflect.DeepEqual(got, want) {
		t.Fatalf("got %+v", got)
	}
}

func TestOptionalWireFieldsAreSentOnlyWhenSet(t *testing.T) {
	version := "1.25"
	tags := wireTags([]Tag{{Type: "language", Value: "go"}, {Type: "language", Value: "go", Version: "1.25"}})
	want := []*memoryv1.Tag{{Type: "language", Value: "go"}, {Type: "language", Value: "go", Version: &version}}
	if len(tags) != 2 || !proto.Equal(tags[0], want[0]) || !proto.Equal(tags[1], want[1]) || tags[0].Version != nil {
		t.Fatalf("tags %v", tags)
	}
	if wireTags(nil) != nil {
		t.Fatal("no tags became an empty slice")
	}
	comment := "why"
	ratings := wireRatings([]FeedbackRating{{Idx: "i", Relevant: true}, {Idx: "j", Correct: true, Comment: "why"}})
	if ratings[0].Comment != nil || !proto.Equal(ratings[1], &memoryv1.FeedbackRating{Idx: "j", Correct: true, Comment: &comment}) {
		t.Fatalf("ratings %v", ratings)
	}
}

func TestAnUnsetSourceIsSentAsAgent(t *testing.T) {
	if wireSource(DataSourceUnspecified) != memoryv1.DataSource_DATA_SOURCE_AGENT {
		t.Fatal("unset")
	}
	if wireSource(DataSourceUser) != memoryv1.DataSource_DATA_SOURCE_USER {
		t.Fatal("user")
	}
}

func TestImportedMemoriesAreCopiedNotShared(t *testing.T) {
	input := []ImportedMemory{{
		Queries:  []string{"q"},
		Insights: []ImportedInsight{{Title: "t", Content: "c"}},
		Tags:     []Tag{{Type: "a", Value: "b"}},
	}}
	got := wireMemories(input)
	want := &memoryv1.ImportedMemory{
		Queries:  []string{"q"},
		Insights: []*memoryv1.ImportedInsight{{Title: "t", Content: "c"}},
		Tags:     []*memoryv1.Tag{{Type: "a", Value: "b"}},
	}
	if len(got) != 1 || !proto.Equal(got[0], want) {
		t.Fatalf("got %v", got)
	}
	got[0].Queries[0] = "changed"
	if input[0].Queries[0] != "q" {
		t.Fatal("the wire copy shares the caller's slice")
	}
}
